src/mpid/ch4/src/ch4_part.c

```diff
@@ -0,0 +1,128 @@
+
+    MPIR_Comm_add_ref(comm);
+    req->comm = comm;
+
+    MPIR_Datatype_add_ref_if_not_builtin(datatype);
+    MPIDI_PART_REQUEST(req, datatype) = datatype;
+
+    MPIDI_PART_REQUEST(req, rank) = rank;
+    MPIDI_PART_REQUEST(req, tag) = tag;
+    MPIDI_PART_REQUEST(req, context_id) = comm->context_id;
+    MPIDI_PART_REQUEST(req, buffer) = buf;
+    MPIDI_PART_REQUEST(req, count) = count;
+    MPIDI_PART_REQUEST(req, active_flag) = 0;
+    req->u.part.partitions = partitions;
+
+    /* Inactive partitioned comm request can be freed by request_free.
+     * Completion cntr increases when request becomes active at start. */
+    MPIR_cc_set(req->cc_ptr, 0);
+
+  fn_exit:
+    MPIR_FUNC_VERBOSE_EXIT(MPID_STATE_MPID_PART_REQ_CREATE);
+    return req;
+  fn_fail:
+    goto fn_exit;
+}
+
+int MPID_Psend_init(void *buf, int partitions, MPI_Count count,
+                    MPI_Datatype datatype, int dest, int tag,
+                    MPIR_Comm * comm, MPIR_Info * info, MPIR_Request ** request)
+{
+    int mpi_errno = MPI_SUCCESS;
+
+    MPIR_FUNC_VERBOSE_STATE_DECL(MPID_STATE_MPID_PSEND_INIT);
+    MPIR_FUNC_VERBOSE_ENTER(MPID_STATE_MPID_PSEND_INIT);
+
+    MPIDI_av_entry_t *av = MPIDIU_comm_rank_to_av(comm, dest);
+
+    MPID_THREAD_CS_ENTER(VCI, MPIDI_VCI(0).lock);
+    MPIDI_workq_vci_progress_unsafe();
+
+    *request = part_req_create((void *) buf, partitions, count, datatype, dest, tag, comm,
+                               MPIR_REQUEST_KIND__PART_SEND);
+#ifdef MPIDI_CH4_DIRECT_NETMOD
+    mpi_errno = MPIDI_NM_mpi_psend_init(buf, partitions, count, datatype, dest, tag, comm,
+                                        info, av, request);
+#else
+    int is_local = MPIDI_av_is_local(av);
+    MPIDI_REQUEST(*request, is_local) = is_local;       /* use at start, pready, parrived */
+
+    if (is_local) {
+        mpi_errno = MPIDI_SHM_mpi_psend_init(buf, partitions, count, datatype, dest, tag,
+                                             comm, info, av, request);
+    } else {
+        mpi_errno = MPIDI_NM_mpi_psend_init(buf, partitions, count, datatype, dest, tag,
+                                            comm, info, av, request);
+    }
+#endif
+    MPIR_ERR_CHECK(mpi_errno);
+
+  fn_exit:
+    MPID_THREAD_CS_EXIT(VCI, MPIDI_VCI(0).lock);
+    MPIR_FUNC_VERBOSE_EXIT(MPID_STATE_MPID_PSEND_INIT);
+    return mpi_errno;
+  fn_fail:
+    goto fn_exit;
+}
+
+int MPID_Precv_init(void *buf, int partitions, MPI_Count count,
+                    MPI_Datatype datatype, int source, int tag,
+                    MPIR_Comm * comm, MPIR_Info * info, MPIR_Request ** request)
+{
+    int mpi_errno = MPI_SUCCESS;
+
+    MPIR_FUNC_VERBOSE_STATE_DECL(MPID_STATE_MPID_PRECV_INIT);
+    MPIR_FUNC_VERBOSE_ENTER(MPID_STATE_MPID_PRECV_INIT);
+
+    MPIDI_av_entry_t *av = MPIDIU_comm_rank_to_av(comm, source);
+
+    MPID_THREAD_CS_ENTER(VCI, MPIDI_VCI(0).lock);
+    MPIDI_workq_vci_progress_unsafe();
```

Comment:
Unless you have specific reason, let's not bother adding the workq progress here. We are not going to add workq for partitioned communication. And workq is being deprecated anyway.